### PR TITLE
Remove JSON schema validation

### LINKED ISSUE #5402: Use protobuf definitions instead of JSON schema for storage data validation
### Feature Request

Remove all JSON schema usage from `api/types` and rely on protobuf schema for validation.
Use https://godoc.org/github.com/gogo/protobuf/jsonpb or something similar to parse storage data to/from in-memory protos.

### Motivation

JSON schema was added when teleport didn't use protobufs yet.
Protobufs already define the schema for data, and it can be used to validate JSON blobs while parsing.
This makes JSON schema redundant. Removing it will reduce the amount of boilerplate needed when working on data storage changes.

Caution: we mix github.com/gogo/protobuf with google.golang.org/protobuf, which may cause surprising behavior. For example, github.com/golang/protobuf/jsonpb is deprecated in favor of google.golang.org/protobuf/encoding/protojson which doesn't work with gogoproto-generated stubs. But github.com/gogo/protobuf/jsonpb from gogoproto doesn't respect JSON struct tags.

### Who's it for?

Teleport team

### PR DESCRIPTION
**Pull request by awly**
Removing JSON schema validation from all resource unmarshalers.

## what JSON schema gets us

Looking at the JSON schema spec and our usage, here are the supposed benefits:
- type validation - make sure incoming data uses the right types for the right fields
- required fields - make sure that mandatory fields are set
- defaulting - set defaults for fields
- documentation - schema definition for our API objects

Note that it does _not_ do:
- fail on unknown fields in data
- fail on a required field with an empty value

## what replaces it

Based on the above, it may seem like JSON schema provides value.
But it's not the case, let's break it down one by one:
- type validation - unmarshaling JSON into a typed Go struct does this
- required fields - only checks that the field was provided, doesn't actually check that a value is set (e.g. `"name": ""` will pass the `required` check)
  - so it's pretty useless for any real validation
  - and we already have a separate place for proper validation - `CheckAndSetDefaults` methods
- defaulting - done in `CheckAndSetDefaults` methods
  - `Version` is the only annoying field, had to add it in a bunch of objects
- documentation - protobuf definitions are the source of truth for our API schema

## the benefits

- performance - schema validation does a few rounds of `json.Marshal/Unmarshal` in addition to actual validation; now we simply skip all that
- maintenance - no need to keep protobuf and JSON schema definitions in sync anymore
- creating new API objects - one error-prone step removed
- lower cognitive load - especially for newer team members
- (future) fewer dependencies - we can _almost_ remove the Go libraries for schema validation (one transient dependency keeping them around)

Fixes https://github.com/gravitational/teleport/issues/5402

### CODE REVIEW

**Review by Joerger — APPROVED** [review]

File: api/types/github.go [hunk 1]
```diff
@@ -161,6 +161,12 @@ func (c *GithubConnectorV3) CheckAndSetDefaults() error {
 	if err := c.Metadata.CheckAndSetDefaults(); err != nil {
 		return trace.Wrap(err)
 	}
+	if c.Version == "" {
+		c.Version = V3
+	}
+	if c.Kind == "" {
+		c.Kind = KindGithubConnector
+	}
```
Comment (andrejtokarcik) [inline api/types/github.go:169]: Shouldn't a similar default be provided for all the resources?

**Review by andrejtokarcik — COMMENTED** [review]
Following up on https://github.com/gravitational/teleport/pull/6825#discussion_r631070478 we could introduce distinct `ResourceHeader{V1,V2,V3,...}` types that would take care of the `Version` field checks automatically.

**Review by andrejtokarcik — APPROVED** [review]

File: api/types/github.go (hunk 1 above)
Comment (Joerger) [reply api/types/github.go:169]: I'm going to make a PR for that and other validation consistency, likely today. Old draft - #5958

Comment (awly) [reply api/types/github.go:169]: Yeah, let's do that separately.
For this PR I want to keep the defaulting behavior as close as possible to what JSON schema did.

**Comment by awly** [comment]
@r0mant @russjones please review

File: lib/auth/auth.go [hunk 2]
```diff
@@ -1077,10 +1077,6 @@ func (a *Server) ExtendWebSession(req WebSessionReq, identity tlsca.Identity) (s
 		return nil, trace.Wrap(err)
 	}
 
-	sess, err = services.ExtendWebSession(sess)
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
```
Comment (r0mant) [inline lib/auth/auth.go:1083]: Did you mean to delete this?

**Review by r0mant — COMMENTED** [review]
In general I don't mind removing jsonschema if it doesn't make things worse. @awly Could you please give a few UX examples on how invalid resources (like, bad indentation, typos, extra fields, etc.) were being handled before vs what kind of errors they're producing now?

File: lib/auth/auth.go (hunk 2 above)
Comment (awly) [reply lib/auth/auth.go:1083]: yeah, see https://github.com/gravitational/teleport/pull/6685/files#diff-6d9bede626eef054444c2f0e375e87693700f2cd99dffa70e9681eebd2f697e8, that function was a no-op

**Comment by awly** [comment]
@r0mant  Using `tctl create -f role.yaml` as an example, here are the outputs for some common validation issues.

With JSON schema:
```sh
$ for fname in (ls role*.yaml); echo "--- $fname"; tctl -c teleport.yaml create -f $fname; end
--- role-incorrect-field-type.yaml
ERROR: error unmarshaling JSON: while decoding JSON: json: cannot unmarshal number into Go struct field Metadata.metadata.name of type string

--- role-missing-kind.yaml
ERROR: creating resources of type "" is not supported

--- role-missing-required-field.yaml
ERROR: failed to validate: name: name is required

--- role-missing-version.yaml
ERROR: role version "" is not supported

--- role-unknown-field.yaml
role 'test-role' has been created
--- role-valid.yaml
role 'test-role' has been created
```

Without JSON schema:
```sh
$ for fname in (ls role*.yaml); echo "--- $fname"; tctl -c teleport.yaml create -f $fname; end
--- role-incorrect-field-type.yaml
ERROR: error unmarshaling JSON: while decoding JSON: json: cannot unmarshal number into Go struct field Metadata.metadata.name of type string

--- role-missing-kind.yaml
ERROR: creating resources of type "" is not supported

--- role-missing-required-field.yaml
ERROR: missing parameter Name

--- role-missing-version.yaml
ERROR: role version "" is not supported

--- role-unknown-field.yaml
role 'test-role' has been created
--- role-valid.yaml
role 'test-role' has been created
```

The only difference is with formatting of errors about required fields.

**Comment by awly** [comment]
@r0mant PTAL

**Review by r0mant — APPROVED** [review]
The only UX issue I see is that if previously the errors were referring to field names as they appear in yaml/json, now they are going to refer to Go struct field names - which may potentially be a bit confusing for the user - especially if there's any mismatch on our side between Go struct field name and its respective yaml key name. I don't know if this UX issue outweighs the benefits of removing json schema or not. Maybe this could be solved by returning more appropriate errors rather than saying something like "field Name is missing" everywhere. I would get @russjones and @klizhentas opinion as well.

**Pull request merged**